src/index.js

```diff
@@ -23,28 +23,51 @@ function setUrl(url, type='push') {
 }
 
 
+function getCurrentLocation() {
+	return (customHistory && customHistory.location) ||
+		(customHistory && customHistory.getCurrentLocation && customHistory.getCurrentLocation()) ||
+		(typeof location!=='undefined' ? location : EMPTY);
+}
+
+
 function getCurrentUrl() {
-	let url;
-	if (customHistory && customHistory.location) {
-		url = customHistory.location;
-	}
-	else if (customHistory && customHistory.getCurrentLocation) {
-		url = customHistory.getCurrentLocation();
-	}
-	else {
-		url = typeof location!=='undefined' ? location : EMPTY;
-	}
+	let url = getCurrentLocation();
 	return `${url.pathname || ''}${url.search || ''}`;
 }
 
 
+const a = typeof document!=='undefined' && document.createElement('a');
+
+// Based on https://tools.ietf.org/html/rfc3986#appendix-B
+const uriRegex = new RegExp('^([^:/?#]+:)?(?://([^/?#]*))?([^?#]*)((?:\\?[^#]*)?)((?:#.*)?)');
+
+/* Resolve URL relative to current location */
+function resolve(url) {
+	let current = getCurrentLocation();
+	if (a) {
+		a.setAttribute('href', url);
+		url = a.href;
+	}
+	let [,protocol,host,pathname,search,hash] = uriRegex.exec(url);
```

Comment:
Could we skip the parse entirely and just do this?

```js
return a.pathname + a.search + a.hash;
```

Reply:
I may have been worrying about server-side (nodejs) rendering where we don't have an `a` to do URL resolution: is that a valid concern?

If not, I'll happily simplify.

Reply:
I think we can probably skip the SSR concern for now - route() should be a no-op there anyway since SSR is single-pass.